### PR TITLE
Navigation: Refactor the PHP render function to make it easier to make changes in the future

### PR DESCRIPTION
**Pull request by scruffian**
## What?
This refactors the PHP render function for the navigation block.

## Why?

The aims are:

1. To make the code easier to grok.
2. To keep variables contained to the context in which they are used
3. To make it easier to change the code in the future.

## How?
Creates a new class called WP_Navigation_Block so that the new functions can be kept private.

## Testing Instructions
Check that the navigation block continues to behave as before.

## Note
This removed the `.is-fallback` CSS class from the navigation block when it's in a fallback state. We don't use this class in core and I can't see a value to it for themes or users.

### CODE REVIEW

File: packages/block-library/src/navigation/index.php
Comment (scruffian) [inline packages/block-library/src/navigation/index.php]: My main concern is that this function is now called several times so its probably slightly less performant.

File: packages/block-library/src/navigation/index.php
Comment (getdave) [inline packages/block-library/src/navigation/index.php]: I think this class might be too generic? WDYT? Is there a danger it becomes a dumping ground for code with no other home?

How about we make it specific to Nav block _rendering_?
```suggestion
class WP_Navigation_Block_Renderer {
```

File: packages/block-library/src/navigation/index.php
Comment (getdave) [inline packages/block-library/src/navigation/index.php]:
```suggestion
		 * This is for backwards compatibility after the `isResponsive` attribute was been removed.
```

Nit: comprehension

File: packages/block-library/src/navigation/index.php
Comment (getdave) [inline packages/block-library/src/navigation/index.php]:
```suggestion
	private static function is_responsive( $attributes ) {
```

Do we need to include the `navigation` element in all the methods? The class already provides the context.

File: packages/block-library/src/navigation/index.php
Comment (getdave) [reply packages/block-library/src/navigation/index.php]: We could exit the loop early and return from the method if a submenu is found. I think `continue` will do that 🤔

Comment (getdave) [reply packages/block-library/src/navigation/index.php]: > now called several times

Why is this different than before?

> probably slightly less performant.

Also this is likely to be marginal. Is it a big concern?

Comment (getdave) [reply packages/block-library/src/navigation/index.php]: ~Let's create an instance variable to cache the result of the function. Then we can check if the instance variable has already been assigned and exit early with that result.~

Actually it's a static class so doing this makes no sense. The class doesn't not represent a block instance.

I'm now wondering if the class should _not_ be static. The instance represents a Navigation and the class is a blue print for hanlding that Nav

**Review by getdave — COMMENTED** [review]
Good PR 👍 

Initial thoughts...

File: packages/block-library/src/navigation/index.php
Comment (getdave) [inline packages/block-library/src/navigation/index.php]: So clean 🏅

File: packages/block-library/src/navigation/index.php
Comment (getdave) [inline packages/block-library/src/navigation/index.php]:
```suggestion
			WP_Navigation_Block::get_wrapper_markup( $attributes, $inner_blocks )
```

Maybe let's use the established convention?

File: packages/block-library/src/navigation/index.php
Comment (getdave) [inline packages/block-library/src/navigation/index.php]: I'm looking at this thinking...

> How is this different from `render`?

😕 Is there a better name?

File: packages/block-library/src/navigation/index.php
Comment (getdave) [inline packages/block-library/src/navigation/index.php]:
```suggestion
	private static function handle_view_script_loading( $block, $attributes, $inner_blocks ) {
```

File: packages/block-library/src/navigation/index.php
Comment (getdave) [inline packages/block-library/src/navigation/index.php]:
```suggestion
		// When adding to this array be mindful of security concerns.
		$nav_element_context = wp_json_encode(
			array(
```

File: packages/block-library/src/navigation/index.php
Comment (getdave) [inline packages/block-library/src/navigation/index.php]:
```suggestion
			$inner_blocks_html = block_core_navigation_add_directives_to_submenu( $w, $attributes );
```

Is this function in the submenu block? If so we should probably comment by way of explanation.

File: packages/block-library/src/navigation/index.php
Comment (scruffian) [reply packages/block-library/src/navigation/index.php]: > Why is this different than before?

Before it was called in the render function and saved in a variable.

> Also this is likely to be marginal. Is it a big concern?

No

File: packages/block-library/src/navigation/index.php
Comment (getdave) [inline packages/block-library/src/navigation/index.php]: This part of the method almost feels that it would be better as a separate method. Why? Because then you could unit test this code around wrapping and accessibility.

I really think we need tests on this part.

File: packages/block-library/src/navigation/index.php
Comment (scruffian) [reply packages/block-library/src/navigation/index.php]: Done in e5c9b18fe4

File: packages/block-library/src/navigation/index.php
Comment (getdave) [inline packages/block-library/src/navigation/index.php]: We should break all conditional branches into their own methods.

- for_unstable_location
- from_navigation post
- from_fallback

File: packages/block-library/src/navigation/index.php
Comment (getdave) [inline packages/block-library/src/navigation/index.php]: We need a comment around this. Or even better a method which handles this code.

**Review by getdave — COMMENTED** [review]
More comments..

File: packages/block-library/src/navigation/index.php
Comment (scruffian) [reply packages/block-library/src/navigation/index.php]: Done in 5cbd5c5f10, bb219fe16b, 44fca51dea, 4c9c4708bd

**Review by getdave — COMMENTED** [review]
The thing I dislike about the PR is that we're refactoring without any test coverage. So the only way we will know if refactoring broke something is when a bug is found 😬 

But given the lack of existing tests I don't see much option. 

I think we should add unit tests for a select few methods at least,

File: packages/block-library/src/navigation/index.php
Comment (scruffian) [reply packages/block-library/src/navigation/index.php]: done in 5157393ba7

File: packages/block-library/src/navigation/index.php
Comment (scruffian) [reply packages/block-library/src/navigation/index.php]: No it's in this file.

File: packages/block-library/src/navigation/index.php
Comment (scruffian) [reply packages/block-library/src/navigation/index.php]: I'm not sure we can separate this part without passing a lot of parameters to it. The logic requires the use of the following variables which are created earlier in the method:

`$container_attributes`, `$class`, `$style`, `$needs_list_item_wrapper`, `$list_item_nav_blocks`.

We could remove the end part which adds the directives if you felt that should be separate.

File: packages/block-library/src/navigation/index.php
Comment (scruffian) [reply packages/block-library/src/navigation/index.php]: Done in [cd2c7fd](https://github.com/WordPress/gutenberg/pull/55605/commits/cd2c7fd5fe441297f3aa8a753b089091d376088e)

File: packages/block-library/src/navigation/index.php
Comment (scruffian) [reply packages/block-library/src/navigation/index.php]: Done in [5514082](https://github.com/WordPress/gutenberg/pull/55605/commits/55140822d4434e88f9292bb29e09796aa03f1d50)

**Review by draganescu — COMMENTED** [review]
~~Aside from the fact that it makes the rendering guts of the block private by encapsulating them in a class, the refactor does not help much IMO with clartiy or anything else. Is it worth it?~~

Dumb comment, I missed that the render function was split, I mistakenly skimmed my trunk copy and assumed all those functions existed just not private. But no, this PR splits the giant render function in smaller bits, and avoids creating a maintainable surface in the process.

File: packages/block-library/src/navigation/index.php
Comment (draganescu) [inline packages/block-library/src/navigation/index.php]: Let's avoid this single char var name. What is `$w`? Maybe we should not pass the object returned by initialising `WP_HTML_Tag_Processor` but instead `$inner_blocks_html ` since processing seems to be the business of `block_core_navigation_add_directives_to_submenu`

The duotone class uses `$tags = new WP_HTML_Tag_Processor( $block_content );` which seems better.

File: packages/block-library/src/navigation/index.php
Comment (draganescu) [inline packages/block-library/src/navigation/index.php]: Maybe this should be bundled witht he two functions at the top that only exist if the plugin is active?

File: lib/load.php
Comment (scruffian) [inline lib/load.php:122]: I don't think this is the right place for this class, but I'm not sure where it should go.

File: packages/block-library/src/navigation/index.php
Comment (scruffian) [reply packages/block-library/src/navigation/index.php]: > Maybe we should not pass the object returned by initialising WP_HTML_Tag_Processor but instead $inner_blocks_html

I agree but this would be changing the signature of that function which might break backwards compatibility?

File: packages/block-library/src/navigation/index.php
Comment (scruffian) [reply packages/block-library/src/navigation/index.php]: done in 70cf09b5ff

File: lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php
Comment (anton-vlasenko) [inline lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php]: All occurrences of
```php
WP_Navigation_Block_Renderer::function_name()
```
in this class should be replaced with
```php
static::function_name()
```
.
The `static` and `self` operators in PHP are built for this purpose and it's a common practice to use them instead of the class name when calling static functions of the same class.
```suggestion
		$has_submenus       = static::has_submenus( $inner_blocks );
```

Applies elsewhere.

**Review by anton-vlasenko — COMMENTED** [review]
Thank you for the PR. This is just my initial review. I might add a follow-up review later, as I am AFK today. Thank you.

File: lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php
Comment (scruffian) [reply lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php]: thanks, done in d3c876b84c

File: packages/block-library/src/navigation/index.php
Comment (getdave) [reply packages/block-library/src/navigation/index.php]: The bit that needs test coverage is the code which determines whether to wrap with lists or not. It's prone to error.

At least two of the parameters you mention could be extracted to static members on the class rather than being local to the method. That leaves 3 params which doesn't feel excessive.

I think it would be worth it to slim down the parent method which is currently doing too much and also to provide more confidence about this particularly unusual piece of code.

File: packages/block-library/src/navigation/index.php
Comment (getdave) [reply packages/block-library/src/navigation/index.php]: I'm starting to wonder if we'll need a follow up PR to move all directives stuff to its own class.

File: lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php
Comment (getdave) [inline lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php]:
```suggestion
	private static $list_item_nav_blocks = array(
```

File: lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php
Comment (getdave) [inline lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php]:
```suggestion
	private static $needs_list_item_wrapper = array(
```

File: packages/block-library/src/navigation/index.php
Comment (scruffian) [reply packages/block-library/src/navigation/index.php]: I added a new function that just handles this logic in 103c27ab2d

File: lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php
Comment (getdave) [inline lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php]: Can we extract the conditional so the concept of "block needs a wrapper" is stored in a well renamed variable.

That means I don't have to study the conditional to understand the code.

File: lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php
Comment (getdave) [inline lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php]: I don't think this var is well named. What does it mean?

PS: I know you didn't come up with the name but can we improve it?

File: lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php
Comment (getdave) [inline lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php]:
```suggestion
				$has_submenus = true;
				break;
			}
```

Could we exit the loop early once we've determine there are submenus?

File: lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php
Comment (getdave) [inline lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php:173]: Why do we do this `$tags` work outside of the function?

File: lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php
Comment (getdave) [inline lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php]: Method here? `get_unique_navigation_name`?

File: lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php
Comment (scruffian) [reply lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php]: done in 295bd0370c

File: lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php
Comment (scruffian) [reply lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php]: done in 939c50ff52

File: lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php
Comment (scruffian) [reply lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php:173]: `block_core_navigation_add_directives_to_submenu` is a pre-existing function. If we move $tags inside it we'll need to change the signature. Is that safe to do or should we be concerned that someone might be using `block_core_navigation_add_directives_to_submenu`?

Comment (getdave) [reply lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php:173]: Let's leave it. Thank you!

File: lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php
Comment (scruffian) [reply lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php]: Done in b16a3bd552

File: lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php
Comment (scruffian) [reply lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php]: renamed to `nav_blocks_wrapped_in_list_item` in aceed2e815

File: phpunit/class-wp-navigation-block-renderer-test.php
Comment (getdave) [inline phpunit/class-wp-navigation-block-renderer-test.php:45]:
```suggestion
	public function test_gutenberg_specialized_blocks_are_enclosed_in_li_tags() {
```

File: phpunit/class-wp-navigation-block-renderer-test.php
Comment (getdave) [inline phpunit/class-wp-navigation-block-renderer-test.php]:
```suggestion
	public function test_gutenberg_default_block_is_enclosed_in_li_tags() {
```

File: phpunit/class-wp-navigation-block-renderer-test.php
Comment (getdave) [inline phpunit/class-wp-navigation-block-renderer-test.php]:
```suggestion
		// We are testing the site title block because we manually add list items around it.
```

File: lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php
Comment (getdave) [inline lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php:25]: Why is this in _both_ of the lists? How can it both be wrapped and need a wrapper?

File: phpunit/class-wp-navigation-block-renderer-test.php
Comment (MaggieCabrera) [inline phpunit/class-wp-navigation-block-renderer-test.php]:
```suggestion
		// Invoke the private method.
```

File: phpunit/class-wp-navigation-block-renderer-test.php
Comment (MaggieCabrera) [inline phpunit/class-wp-navigation-block-renderer-test.php]:
```suggestion
		// We are testing the site title block because we manually add list items around it.
```

File: phpunit/class-wp-navigation-block-renderer-test.php
Comment (MaggieCabrera) [inline phpunit/class-wp-navigation-block-renderer-test.php]:
```suggestion
		// Invoke the private method.
```

File: lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php
Comment (scruffian) [reply lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php:25]: It needs a wrapper, so once it has been wrapped, it is wrapped! 🤣

**Review by MaggieCabrera — APPROVED** [review]
Besides the unit tests that we included on this PR, I've tested this comparing the markup generated for a navigation block that has: 
- a submenu
- a site title block
- a logo block
- social icons

The markup generated on trunk and with this PR is exactly the same except for the `is-fallback` class that has been removed. I think this is good to go!

**Comment by scruffian** [comment]
I decided to merge this so that we don't end up with messy rebases when changes get made to the navigation render function. I expect there are still opportunities to improve this code though, so please leave more reviews and I'll follow up with more improvements. I'll also submit a core patch now so that we can get reviews from core committers.

**Comment by scruffian** [comment]
I opened a patch in core for this here: https://core.trac.wordpress.org/ticket/59867

**Comment by Mamaduka** [comment]
What's the strategy for patching bugs or making enhancements for this class after it ships in the core?

I'm asking because we started discouraging `Gutenberg_` prefixed classes and functions in the `block-library` package, as they can easily slip in during package sync and break the core.

**Comment by tellthemachines** [comment]
> What's the strategy for patching bugs or making enhancements for this class after it ships in the core?

What we've done previously with e.g. the [theme json class](https://github.com/WordPress/gutenberg/blob/trunk/lib/class-wp-theme-json-gutenberg.php#L0-L1) is make a copy of the whole class in Gutenberg, namespaced so it doesn't conflict with the core one, and use it in the plugin instead of the core class. If we're using the class directly in the block library PHP files that makes things a little more complicated 😅 but I'm sure we can come up with some auto-renaming script that will switch the core class with the gutenberg one at build time.

File: packages/block-library/src/navigation/index.php
Comment (tellthemachines) [inline packages/block-library/src/navigation/index.php:82]: I'm seeing a warning `Deprecated: Function gutenberg_block_core_navigation_parse_blocks_from_menu_items is deprecated since version 6.3.0! Use WP_Navigation_Fallback::parse_blocks_from_menu_items instead. in /var/www/html/wp-includes/functions.php on line 6031` on trunk when using a hybrid theme with some legacy Navigation blocks in it. Would it be possible to use `WP_Navigation_Fallback::parse_blocks_from_menu_items` here?

File: phpunit/class-wp-navigation-block-renderer-test.php
Comment (getdave) [inline phpunit/class-wp-navigation-block-renderer-test.php:9]: @scruffian I think we should move this test into the `phpunit/blocks` directory. It is specific to a specific block and thus should be managed within the Gutenberg repo. 

See https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/back-merging-to-wp-core.md#ignored-directoriesfiles

**Comment by getdave** [comment]
Noting that the renderer file code was moved into the block-library package in https://github.com/WordPress/gutenberg/pull/57979 and thus should be automatically handled by the packages sync process.

However, [the phpunit test should be moved into `phpunit/blocks`](https://github.com/WordPress/gutenberg/pull/55605#discussion_r1466193262) to avoid the need to backport to WP Core.

File: phpunit/class-wp-navigation-block-renderer-test.php
Comment (scruffian) [reply phpunit/class-wp-navigation-block-renderer-test.php:9]: Done in https://github.com/WordPress/gutenberg/pull/58247

**Comment by oandregal** [comment]
👋 We need a follow-up to address some performance issues. I've prepared https://github.com/WordPress/gutenberg/pull/58506 to demonstrate.

**Pull request merged**
